babel/messages/frontend.py

```diff
@@ -1053,6 +1025,68 @@ def parse_mapping(fileobj, filename=None):
     return method_map, options_map
 
 
+def _parse_config_object(config: dict, *, filename="(unknown)"):
+    extractors = {}
+    method_map = []
+    options_map = {}
+
+    extractors_read = config.get("extractors", {})
+    if not isinstance(extractors_read, dict):
+        raise ValueError(f"{filename}: extractors: Expected a dictionary, got {type(extractors_read)!r}")
+    for method, callable_spec in extractors_read.items():
+        if not isinstance(method, str):
+            raise ValueError(f"{filename}: extractors: Extraction method must be a string, got {method!r}")
+        if not isinstance(callable_spec, str):
+            raise ValueError(f"{filename}: extractors: Callable specification must be a string, got {callable_spec!r}")
+        extractors[method] = callable_spec
+
+
+    mappings_read = config.get("mappings") or []
+    if not isinstance(mappings_read, list):
+        raise ValueError(f"{filename}: mappings: Expected a list, got {type(mappings_read)!r}")
+    for idx, entry in enumerate(mappings_read):
+        if not isinstance(entry, dict):
+            raise ValueError(f"{filename}: mappings[{idx}]: Expected a dictionary, got {type(entry)!r}")
+        entry = entry.copy()
+        method = entry.pop("method", None)
+        pattern = entry.pop("pattern", None)
+        if not isinstance(method, str):
+            raise ValueError(f"{filename}: mappings[{idx}]: 'method' must be a string, got {method!r}")
+        if not isinstance(pattern, str):
+            raise ValueError(f"{filename}: mappings[{idx}]: 'pattern' must be a string, got {pattern!r}")
+        method = extractors.get(method, method)  # Map the extractor name to the callable now
+        method_map.append((pattern, method))
+        options_map[pattern] = entry
+
+    return method_map, options_map
+
+
+def parse_mapping_toml(fileobj: BinaryIO, filename: str = "(unknown)"):
+    """Parse an extraction method mapping from a binary file-like object.
+
+    :param fileobj: a readable binary file-like object containing the configuration TOML to parse
+    :param filename: the name of the file being parsed, for error messages
+    :see: `extract_from_directory`
+    """
+    try:
+        import tomli as tomllib
+    except ImportError:
+        try:
+            import tomllib
+        except ImportError as ie:
+            raise ImportError("tomli or tomllib is required to parse TOML files") from ie
+
+    parsed_data = tomllib.load(fileobj)
+    babel_data = None
+    if "babel" in parsed_data:
+        babel_data = parsed_data["babel"]
+    elif "tool" in parsed_data and isinstance(parsed_data["tool"], dict):
+        babel_data = parsed_data["tool"].get("babel")
```

Comment:
I find supporting both `[babel]` and `[tool.babel]` to be surprising, and not in line with usual `pyproject.toml` usage. To still support both `babel.toml` and `pyproject.toml` configs, the expected path could depend on the file name?

Reply:
Yeah, I guess that's a little surprising and I think I was going from where the aita is the matalin here. I referred to how `ruff` does this thing, and they do have separate code paths for dealing with `ruff.toml` and `pyproject.toml`, so I suppose we can just as well too.

That said, a `pyproject.toml` with a top-level `babel` key would probably fail [`validate-pyproject`](https://pypi.org/project/validate-pyproject/) but you're right - better not be sloppy now that we have the chance not to be.